csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,315 @@
+    GroupIdMap computed_ids;
+
+    // Look up each loop IterDomain in tv and assert that it is in
+    // scheduled_val_groups_, then map it.
+    for (IterDomain* loop_id : tv->getLoopDomain()) {
+      ValGroup vg = graph_->toGroup(loop_id);
+      // NOTE: this check guarantees that we will only ever _append_ transforms
+      // to the loop domain of tv. If instead we searched for producer IDs in
+      // the root->loop path of tv, then we could wind up silently erasing
+      // some pre-existing transforms.
+      NVF_CHECK(
+          scheduled_val_groups_.count(vg) > 0,
+          "Found loop IterDomain ",
+          loop_id->toString(),
+          " that is not described by abstract tensor.");
+      computed_ids.emplace(vg, loop_id);
+    }
+
+    return computed_ids;
+  }
+
+  IterDomain* replayAbstractId(AbstractId abs_id, GroupIdMap& computed_ids) {
+    const ValGroup& g = abstractIdToValGroup(abs_id);
+
+    // Uncomputable ValGroups are those that we cannot compute from tv's root.
+    // When we detect that any inputs of an ExprGroup are in these groups, we
+    // know that we cannot compute that ExprGroup so we should skip it when
+    // scheduling tv.
+    //
+    // For example, suppose we had
+    //   tv:
+    //     logical/loop: iS0{i0} iS1{i1} iS2{i2}
+    //
+    //   graph:
+    //     ValGroup 0: iS0
+    //     ValGroup 1: iS1
+    //     ValGroup 2: iS2 iS6
+    //     ValGroup 3: iS3
+    //     ValGroup 4: iS4
+    //     ValGroup 5: iS5
+    //     ValGroup 6: iS7
+    //     ValGroup 7: iS8
+    //     ExprGroup 0:
+    //       iS3 = merge(iS0, iS1)
+    //     ExprGroup 1:
+    //       iS4, iS5 = split(iS3, 16)
+    //     ExprGroup 2:
+    //       iS8 = merge(iS6, iS7)
+    //
+    //   abstract_tensor_.domain: ValGroup 4, ValGroup 5, ValGroup 7
+    //
+    //      VG0   VG1
+    //        \   /
+    //         EG0
+    //          |
+    //         VG3       VG2   VG6
+    //          |          \   /
+    //         EG1          EG2
+    //        /   \          |
+    //      VG4   VG5       VG7
+    //
+    // In this case, tv has loop domains in ValGroups 0, 1, and 2. IterDomains
+    // iS6, iS7, and iS8 might be associated to another tensor in the fusion.
+    //
+    // ValGroups 4 and 5 are computable since those ValGroups are produced by
+    // ExprGroup 1 which itself produced by ExprGroup 0, and tv includes iS0 and
+    // iS1.
+    //
+    // However, ValGroup 7 is not computable. It is produced by ExprGroup 2
+    // whose input ValGroups are 2 and 6. ValGroup 2 is computable since iS2 is
+    // in tv, however there is no IterDomain in tv that can be used to represent
+    // ValGroup 6 which also has no producer ValGroups.
+
+    // Any computable ValGroup should have an entry in computed_ids. If we prove
+    // that the ValGroup is not computable, we insert it here.
+    std::unordered_set<ValGroup> uncomputable_groups;
```

Comment:
Please also mention like below:
```
// In the case of the above example, VG6 and VG7 would be included.
```

Reply:
Good idea. I'll leave a comment that VG6 would be included here. In fact we don't mark VG7 as uncomputable here; since VG2 is computable even though VG6 is not, we just use the representative from VG2 to represent VG7 i.e. if there is nothing to merge with we just skip merging and use the existing input ID.

Reply:
> we just use the representative from VG2 to represent VG7 i.e. if there is nothing to merge with we just skip merging and use the existing input ID

Can you elaborate why?

Reply:
For example if we have
```
tv2[ iS1{i0} bS2{1} ] =  broadcast( tv0[ iS0{i0} ] )
tv3[ iS3{i0} iS4{i1} ] = mul(tv1, tv2)
tv3->merge(0)
```
Now if we propagate from tv3, this should leave `iS0` in place in `tv0`. This is the type of thing I was trying to support while still applying the same `AbstractTensor` all tensors.

Reply:
We only do this forwarding for certain patterns of broadcast domains. See `ForwardingInfo`.

Reply:
Is this forwarding behavior necessary for your matmul case? If not, we could revisit later. We may be able to remove that eventually with #2459.

Reply:
I am not sure I really need it actually.

Specifically for scheduling matmul, we need to do the following:

1. Cache epilogue inputs/outputs
2. cache operands for smem and register loads
3. Apply basic tiling to mma_result (i.e. merge dimensions with same roles, reorder so that roles are in canonical order). This results in something like [B, Mo, No, Ko, Mi, Ni, Ki]
4. Split Ko by splitk_factor: [B, Mo, No, Kf, Kg, Mi, Ni, Ki]
5. rFactor({Kf, Ki}) to create splitk_sum
6. Cache mma_result to smem for smem_epilogue
7. Schedule all tensors. This might involve multiple `applyAbstractTransforms` and `parallelizeAllLike` but it should not introduce new tensors.

Steps 1, 2, 5, and 6 create new tensors. Step 5 must happen after some transforms have been applied. With the current machinery I have steps 1-6 and can progress to step 7 which involves a bunch of substeps.

After I have scheduled `mma_result` in step 3, I identify the Ko and Ki dimensions using its loop `ValGroup`s returned by `graph.toGroup(id)`. After step 5, I rebuild the `IdModel` since there are new tensors, and I update the `AbstractTensor` used in steps 3/4 to point to the new ValGroups/ValGraph. At that point, I hope to be able to infer the role of each loop domain in every tensor in the fusion for step 7.

In step 7, we'll need to at times apply the same transforms to pre- and post-broadcast tensors, for example if we are applying a 1D or 0D bias in the epilogue, or when scheduling the prologue tensors that broadcast a missing M or N dimension on either operand. In those cases, it seemed more convenient to just skip merging the missing dimensions than to manually build a different AbstractTensor for each pattern found in the fusion (not that that couldn't also be done if we need to).

Reply:
> Is this forwarding behavior necessary for your matmul case? If not, we could revisit later. We may be able to remove that eventually with #2459.

I think we can work around it, at the expense of more book-keeping. Currently, I schedule a single `AbstractTensor` to represent tiling then I can apply that to every tensor in the matmul fusion. If we disable forwarding, then we will need a different `AbstractTensor` for each group of tensors with/without some axes to avoid an error.

How about this: we disable automatic forwarding and instead move it to a utility like this:
```c++
//! Forward missing axes in merge and swizzle expressions (i.e. ignore the expression
//! and keep the axis intact). Axis presence is determined based on the provided
//! TensorView. All AbstractIds in abten must be of type ValGroupAndItsGraph.
AbstractTensor forwardMissingAxes(const AbstractTensor& abten, TensorView* tv);
```
This way the behavior is explicit and the scheduler is in charge of guaranteeing this is the desired behavior.

Reply:
Sounds reasonable